backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt

```diff
@@ -0,0 +1,281 @@
+        return contestData.toApi() to allRuns
+    }
+
+    private fun parseProblemsInfo(): List<ProblemInfo> {
+        val doc: Document
+        runBlocking {
+            doc = xmlLoader.loadOnce()
+        }
+
+        val config = doc.child(0)
+        config.children().forEach {
+            if ("problems" == it.tagName()) {
+                return it.children().map { element ->
+                    ProblemInfo(
+                        element.attr("short_name"),
+                        element.attr("short_name"),
+                        null
+                    )
+                }
+            }
+        }
+
+        logger.error("There is no <problems> tag in external XML log")
+        return emptyList()
+    }
+
+    private fun parseTeamsInfo(problemsNumber: Int): List<EjudgeTeamInfo> {
+        val doc: Document
+        runBlocking {
+            doc = xmlLoader.loadOnce()
+        }
+
+        val config = doc.child(0)
+        config.children().forEach {
+            if ("users" == it.tagName()) {
+                return it.children().withIndex().map { (index, participant) ->
+                    val participantName = participant.attr("name")
+                    val alias = participant.attr("id")
+                    val groups = listOf<String>()
+                    val medias = mutableMapOf<MediaType, String>()
+                    EjudgeTeamInfo(
+                        TeamInfo(
+                            index,
+                            participantName,
+                            participantName,
+                            alias,
+                            groups,
+                            participantName,
+                            medias
+                        ),
+                        problemsNumber
+                    )
+                }
+            }
+        }
+
+        logger.error("There is no <users> tag in external XML log")
+        return emptyList()
+    }
+
+    private fun parseEjudgeTime(time: String): Instant {
+        val formattedTime = time
+            .replace("/", "-")
+            .replace(" ", "T")
+        return guessDatetimeFormat(formattedTime)
+    }
+
+    private fun parseContestInfo(element: Element, onRunChanges: (RunInfo) -> Unit) {
+        val dur = element.attr("duration").toLong()
+        val startTime = parseEjudgeTime(element.attr("start_time"))
+        val endTime = startTime.plus(dur.seconds)
+        val currentTime = parseEjudgeTime(element.attr("current_time"))
+
+        val status: ContestStatus = if (currentTime >= endTime) {
+            ContestStatus.OVER
+        } else if (currentTime < startTime) {
+            ContestStatus.BEFORE
+        } else {
+            ContestStatus.RUNNING
+        }
+
+        if (status == ContestStatus.RUNNING && contestData.status !== ContestStatus.RUNNING) {
+            contestData.startTime = startTime
+        }
+        contestData.status = status
+        contestData.contestTime = Duration.between(
```

Comment:
what is wrong with currentTime - startTime? As I remember it returns kotlin.time.Duration?